### PR TITLE
fix(sql): split_part returning old data when no match is found

### PR DESCRIPTION
**Pull request by nwoolmer**
Closes https://github.com/questdb/questdb/issues/4489

Previously, the function could early return in the non-matching case without clearing the sink.

~The fix is to clear it at the top of the function.~

The fix is to split the impls so that the local sinks are cleared, and the remote sink not cleared, refactoring the original end-of-function check.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartVarcharFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartCharFunctionFactory.java|2|2|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartCharFunctionFactory.java [hunk 1]
```diff
@@ -148,6 +148,10 @@ public void init(SymbolTableSource symbolTableSource, SqlExecutionContext execut
 
         @Nullable
         private <S extends Utf16Sink> S getStr0(Record rec, S sink, boolean clearSink) {
+            if (clearSink && sink instanceof Mutable) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartCharFunctionFactory.java:151]: we must not have instanceOf on the hot path

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 13 / 17 (76.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartCharFunctionFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartVarcharFunctionFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java|5|6|83.33%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 13 / 17 (76.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartCharFunctionFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartVarcharFunctionFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java|5|6|83.33%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 52 / 74 (70.27%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java|22|33|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartCharFunctionFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/SplitPartVarcharFunctionFactory.java|26|35|74.29%|

**Pull request merged**